sled-agent/config-reconciler/src/reconciler_task.rs

```diff
@@ -127,21 +147,292 @@ pub enum ReconcilerTaskStatus {
+            .await;
+
+        // Ensure all the datasets we want exist.
+        let datasets = match self
+            .dataset_task
+            .datasets_ensure(sled_config.datasets.clone())
+            .await
+        {
+            Ok(result) => result,
+            Err(err) => {
+                warn!(
+                    self.log, "failed to contact dataset task";
+                    InlineErrorChain::new(&err),
+                );
+                // If we can't contact the dataset task, reuse the result from
+                // our previous attempt. This should still be correct (until we
+                // start deleting datasets, at which point we'll need a more
+                // holistic tracker for dataset status like we already have for
+                // disks and zones).
+                self.reconciler_result_tx
+                    .borrow()
+                    .latest_result
+                    .as_ref()
+                    .map(|inner| inner.datasets.clone())
+                    .unwrap_or_else(DatasetEnsureResult::default)
+            }
+        };
+
+        // Collect the current timesync status (needed to start any new zones,
+        // and also we want to report it as part of each reconciler result).
+        let timesync_status = self.zones.check_timesync().await;
+
+        // We conservatively refuse to start any new zones if any zones have
+        // failed to shut down cleanly. This could be more precise, but we want
+        // to avoid wandering into some really weird cases, such as:
+        //
+        // * Multiple NTP zones active concurrently
+        // * Multiple Crucible zones trying to manage the same zpool
+        //
+        // which could happen if we're upgrading a zone but failed to shut down
+        // the old instance.
+        match zone_shutdown_result {
+            Ok(()) => {
+                let currently_managed_zpools =
+                    self.external_disks.currently_managed_zpools();
+                self.zones
+                    .start_zones_if_needed(
+                        &sled_config.zones,
+                        sled_agent_facilities,
+                        timesync_status.is_synchronized(),
+                        &currently_managed_zpools,
+                        &self.log,
+                    )
+                    .await;
+            }
+            Err(nfailed) => {
+                warn!(
+                    self.log,
+                    "skipping attempt to start new zones; \
+                     {nfailed} zones failed to shut down cleanly"
+                );
+            }
+        }
+
+        // We'll retry even if there have been no config changes if (a) time
+        // isn't sync'd yet or (b) any of our disk/dataset/zone attempts failed
+        // with a retryable error.
+        let result = if !timesync_status.is_synchronized()
+            || self.external_disks.has_retryable_error()
+            || self.zones.has_retryable_error()
```

Comment:
Is this always true when `zone_shutdown_result` is an error?

(That also seems like a case we should retry)

Reply:
Yep, as you noted below.